### PR TITLE
Add Build Matrix for CI

### PR DESCRIPTION
**Pull request by EricGustin**
This PR adds various dimensions to the CI build matrix for SmartSim. The build matrix now uses MacOS & Ubuntu, GNU8, RedisAI 1.2.3 & 1.2.5, and Python3.7-3.9. The build matrix excludes building with RedisAI 1.2.5 when on MacOS.

Effort and research was conducted on whether the CI can be ran with the Intel compiler. There is a follow up ticket on these findings. 

In addition to the CI build matrix, there were modifications to `_core/builder.py` by @Spartee that removed the usage of `shell=True` and replaced it with `shell=False` as the default behavior for running commands.

### CODE REVIEW

File: .github/workflows/build_wheels.yml [hunk 1]
```diff
@@ -1,4 +1,4 @@
-name: run-tests-local
+name: build-wheels
```
Comment (Spartee) [inline .github/workflows/build_wheels.yml:1]: Seems like this was a mistake?

File: .github/workflows/run_tests.yml [hunk 2]
```diff
@@ -0,0 +1,70 @@
...
+          sudo add-apt-repository ppa:ubuntu-toolchain-r/test
+          sudo apt-get update
+          sudo apt-get install -y build-essential
+          sudo apt-get install -y wget
+
+      - name: Install GNU make for MacOS
+        if: "contains( matrix.os, 'macos' )"
+        run: brew install make || true
+
+      - name: Install SmartSim on MacOS
```
Comment (Spartee) [inline .github/workflows/run_tests.yml:49]: we can probably collapse the macos and ubuntu stages here into a single run. This decreases the number of docker layers created by the CI and shoul dmake things faster.

File: .github/workflows/build_wheels.yml (hunk 1 above)
Comment (Spartee) [reply .github/workflows/build_wheels.yml:1]: We should still be testing the wheel builds.

Comment (Spartee) [reply .github/workflows/build_wheels.yml:1]: Another thought here, 

we should add
```bash
export SMARTSIM_SUFFIX=develop
```
prior to the wheel build so the wheels are named with the dirty version.

**Review by Spartee — CHANGES_REQUESTED** [review]

File: .github/workflows/build_wheels.yml (hunk 1 above)
Comment (EricGustin) [reply .github/workflows/build_wheels.yml:1]: We are still testing the wheel builds. Wheel builds occur in `build_wheels.yml` and the extended build matrix which will eventually contain code coverage reporting is in `run_tests.yml`.

I added 
```yml
env:
    SMARTSIM_SUFFIX: "develop"
```
to the wheel builds in my most recent push.

**Review by Spartee — APPROVED** [review]
LGTM. Glad we discussed this.

**Pull request merged**